vector/src/main/java/im/vector/app/features/pin/lockscreen/crypto/migrations/MissingSystemKeyMigrator.kt

```diff
@@ -41,14 +39,15 @@ class MissingSystemKeyMigrator @Inject constructor(
      * If user had biometric auth enabled, ensure system key exists, creating one if needed.
      */
     @SuppressLint("NewApi")
-    fun migrate() {
+    fun migrateIfNeeded() {
         if (buildVersionSdkIntProvider.get() >= Build.VERSION_CODES.M && vectorPreferences.useBiometricsToUnlock()) {
-            try {
-                keystoreCryptoFactory.provide(systemKeyAlias, true).ensureKey()
-            } catch (e: KeyPermanentlyInvalidatedException) {
-                Timber.e("Could not automatically create biometric key because it was invalidated.", e)
-            } catch (e: UserNotAuthenticatedException) {
-                Timber.e("Could not automatically create biometric key because there are no enrolled biometric authenticators.", e)
+            val systemKeyStoreCrypto = keystoreCryptoFactory.provide(systemKeyAlias, true)
+            runCatching {
+                systemKeyStoreCrypto.ensureKey()
+            }.onFailure { e ->
+                Timber.e(e, "Could not automatically create biometric key. Biometric authentication will be disabled.")
+                systemKeyStoreCrypto.deleteKey()
+                vectorPreferences.setUseBiometricToUnlock(false)
```

Comment:
does this mean we've fallen back to prompting for a pin?

Reply:
Yes. It's not a great solution, but I can see no alternative. If the migration fails, that's because the key is no longer valid and biometric auth won't work anyway even if the key exists.